packages/subgraph-service/contracts/libraries/IndexingAgreement.sol

```diff
@@ -0,0 +1,519 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.27;
+
+import { IGraphPayments } from "@graphprotocol/horizon/contracts/interfaces/IGraphPayments.sol";
+import { IRecurringCollector } from "@graphprotocol/horizon/contracts/interfaces/IRecurringCollector.sol";
+import { GraphDirectory } from "@graphprotocol/horizon/contracts/utilities/GraphDirectory.sol";
+
+import { SubgraphService } from "../SubgraphService.sol";
+import { Directory } from "../utilities/Directory.sol";
+import { Allocation } from "./Allocation.sol";
+import { SubgraphServiceLib } from "./SubgraphServiceLib.sol";
+import { Decoder } from "./Decoder.sol";
+
+library IndexingAgreement {
+    using IndexingAgreement for Manager;
+    using Allocation for mapping(address => Allocation.State);
+    using SubgraphServiceLib for mapping(address => Allocation.State);
+
+    /// @notice Versions of Indexing Agreement Metadata
+    enum IndexingAgreementVersion {
+        V1
+    }
+
+    /**
+     * @notice Indexer Agreement Data
+     * @param allocationId The allocation ID
+     * @param version The indexing agreement version
+     */
+    struct State {
+        address allocationId;
+        IndexingAgreementVersion version;
+    }
+
+    struct AgreementWrapper {
+        State agreement;
+        IRecurringCollector.AgreementData collectorAgreement;
+    }
+
+    /**
+     * @notice Accept Indexing Agreement metadata
+     * @param subgraphDeploymentId The subgraph deployment ID
+     * @param version The indexing agreement version
+     * @param terms The indexing agreement terms
+     */
+    struct AcceptIndexingAgreementMetadata {
+        bytes32 subgraphDeploymentId;
+        IndexingAgreementVersion version;
+        bytes terms;
+    }
+
+    /**
+     * @notice Update Indexing Agreement metadata
+     * @param version The indexing agreement version
+     * @param terms The indexing agreement terms
+     */
+    struct UpdateIndexingAgreementMetadata {
+        IndexingAgreementVersion version;
+        bytes terms;
+    }
+
+    /**
+     * @notice Indexing Agreement Terms (Version 1)
+     * @param tokensPerSecond The amount of tokens per second
+     * @param tokensPerEntityPerSecond The amount of tokens per entity per second
+     */
+    struct IndexingAgreementTermsV1 {
+        uint256 tokensPerSecond;
+        uint256 tokensPerEntityPerSecond;
+    }
+
+    struct CollectParams {
+        bytes16 agreementId;
+        uint256 currentEpoch;
+        bytes data;
+    }
+
+    /// @custom:storage-location erc7201:graphprotocol.subgraph-service.storage.Manager.IndexingAgreement
+    struct Manager {
+        mapping(bytes16 => State) agreements;
+        mapping(bytes16 agreementId => IndexingAgreementTermsV1 data) termsV1;
+        mapping(address allocationId => bytes16 agreementId) allocationToActiveAgreementId;
+    }
```

Comment:
It's not immediately obvious what this `Manager` is for, I would consider renaming to `StorageManager` and `_getStorageManager()`